src/evm/protocol/uniswap_v4/state.rs

```diff
@@ -1,26 +1,51 @@
-use alloy_primitives::U256;
+#![allow(dead_code)] //TODO: remove when used
 
-use crate::evm::protocol::utils::uniswap::tick_list::{TickInfo, TickList};
+use alloy_primitives::{Sign, I256, U256};
+
+use crate::{
+    evm::protocol::{
+        safe_math::{safe_add_u256, safe_sub_u256},
+        u256_num::u256_to_biguint,
+        utils::uniswap::{
+            liquidity_math, swap_math,
+            tick_list::{TickInfo, TickList, TickListErrorKind},
+            tick_math::{
+                get_sqrt_ratio_at_tick, get_tick_at_sqrt_ratio, MAX_SQRT_RATIO, MAX_TICK,
+                MIN_SQRT_RATIO, MIN_TICK,
+            },
+            StepComputation, SwapResults, SwapState,
+        },
+    },
+    protocol::{errors::SimulationError, models::GetAmountOutResult, state::ProtocolSim},
+};
 
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub struct UniswapV4State {
     liquidity: u128,
     sqrt_price: U256,
-    lp_fee: i32,
-    protocol_fees: UniswapV4ProtocolFees,
+    fees: UniswapV4Fees,
     tick: i32,
     ticks: TickList,
 }
 
 #[derive(Clone, Debug, PartialEq, Eq)]
-pub struct UniswapV4ProtocolFees {
-    zero2one: i32,
-    one2zero: i32,
+pub struct UniswapV4Fees {
+    // Protocol fees in the zero for one direction
+    zero_for_one: u32,
+    // Protocol fees in the one for zero direction
+    one_for_zero: u32,
+    // Liquidity providers fees
+    lp_fee: u32,
 }
 
-impl UniswapV4ProtocolFees {
-    pub fn new(zero2one: i32, one2zero: i32) -> Self {
-        Self { zero2one, one2zero }
+impl UniswapV4Fees {
+    pub fn new(zero_for_one: u32, one_for_zero: u32, lp_fee: u32) -> Self {
+        Self { zero_for_one, one_for_zero, lp_fee }
+    }
+
+    fn calculate_swap_fees_pips(&self, zero_for_one: bool) -> u32 {
```

Comment:
This can be considered as shallow, and we could probably have this directly in the functions body. But I decided to do it like this to stay consistent with how it's done in their smart contract

Reply:
```
    // The protocol fee is taken from the input amount first and then the LP fee is taken from the remaining
    // The swap fee is capped at 100%
    // Equivalent to protocolFee + lpFee(1_000_000 - protocolFee) / 1_000_000 (rounded up)
    /// @dev here `self` is just a single direction's protocol fee, not a packed type of 2 protocol fees
    function calculateSwapFee(uint16 self, uint24 lpFee) internal pure returns (uint24 swapFee) {
        // protocolFee + lpFee - (protocolFee * lpFee / 1_000_000)
        assembly ("memory-safe") {
            self := and(self, 0xfff)
            lpFee := and(lpFee, 0xffffff)
            let numerator := mul(self, lpFee)
            swapFee := sub(add(self, lpFee), div(numerator, PIPS_DENOMINATOR))
        }
    }
```

Shouldn't you do the division or am I missing smth?

Reply:
From what I saw we use pips in our maths, I'm not 100% sure how the substreams gives us things yet though, this will need to be confirmed when testing